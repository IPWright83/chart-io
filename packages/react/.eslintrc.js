module.exports = {
    env: {
        browser: true,
        es2021: true,
    },
    extends: ["eslint:recommended", "plugin:react/recommended", "plugin:@typescript-eslint/recommended"],
    overrides: [
        {
            files: "*.unit.tsx?",
            rules: {
                "valid-jsdoc": "off",
            },
        },
    ],
    parser: "@typescript-eslint/parser",
    parserOptions: {
        ecmaVersion: "latest",
        sourceType: "module",
    },
    plugins: ["react", "@typescript-eslint"],
    ignorePatterns: ["jest.config.js", ".eslintrc.js", "**/jest/*.js", "testUtils/*.ts", "*.stories.tsx"],
    rules: {
        "@typescript-eslint/no-explicit-any": "off",
        "valid-jsdoc": [
            "warn",
            {
                requireReturnType: false,
                requireParamType: false,
                requireReturn: false,
            },
        ],
        "@typescript-eslint/ban-ts-comment": "warn",
    },
};
