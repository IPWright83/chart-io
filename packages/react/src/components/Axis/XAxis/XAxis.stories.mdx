<!--- XAxis.stories.mdx --->

import { Meta, Story, Canvas } from "@storybook/addon-docs/blocks";
import LinkTo from "@storybook/addon-links/react";

import { XAxis } from "./XAxis";
import { Chart } from "../../Chart";

<Meta title="Components/Axis/XAxis" component={XAxis} />

export const XAxisTemplate = (args) => (
    <Chart {...args}>
        <XAxis fields={args.fields} />
    </Chart>
);

# Axis Component

This `<XAxis>` component allows you to define a horizontal Axis on a chart. This is a utility component that combines both an `<Axis>` component, along with a `<Scale>` component. It is recommended to use this component where possible to save duplicate configuration for a `<Scale>`.

Please see the <LinkTo kind="Axis-Axis">Axis</LinkTo> component for more information on it's capabilities and configuration.

<Story
    name="XAxis"
    args={{
        data: [{ x: 0 }, { x: 10 }],
        fields: ["x"],
        height: 100,
        width: 800,
    }}
>
    {XAxisTemplate.bind({})}
</Story>

### Props

| Prop        | Type            | Default | Note                                                                                                                                                   |
| ----------- | --------------- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `position`  | `string`        |         | The position of the axis. One of `top`, `bottom`. This will use the appropriate [d3.axis](https://github.com/d3/d3-axis#axisTop) orientation function. |
| `fields`    | `Array<string>` |         | The set of fields that this axis represents. These should correspond to the fields being displayed on this Axis within the data.                       |
| `scaleType` | `string`        | null    | The type of scale being used. Should be one of `linear`, `time`, `power`, `log`, `band`, `point`                                                       |
| `aggregate` | `boolean`       | false   | Whether values in `fields` should be summed when determining the scale                                                                                 |

<br />
