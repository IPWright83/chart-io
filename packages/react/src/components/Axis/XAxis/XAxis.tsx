import PropTypes from "prop-types";
import React from "react";
import { Axis } from "../Axis";
import { XScale } from "../../Scale";

import type { VerticalPosition, Primitive, ScaleType } from "../../../types";

export interface XAxisProps {
    /**
     * The position of the axis [top, bottom]
     */
    position: VerticalPosition;
    /**
     * (Optional) An override of the domain to use with the d3 scale
     */
    domain: [Primitive, Primitive];
    /**
     * The keys of the fields that will share this scale
     */
    fields: string[];
    /**
     * (Optional) Force the type of d3 scale to use - https://github.com/d3/d3-scale
     */
    scaleType: ScaleType;
    /**
     * Whether this scale is an aggregate (of multiple y values)
     */
    aggregate: boolean;
    /**
     * https://github.com/d3/d3-axis#axis_tickSizeInner
     */
    tickSizeInner: number;
    /**
     * https://github.com/d3/d3-axis#axis_tickSizeOuter
     */
    tickSizeOuter: number;
    /**
     * https://github.com/d3/d3-axis#axis_tickPadding
     */
    tickPadding: number;
    /**
     * Should gridlines be drawn?
     */
    showGridlines: boolean;
    /**
     * A title for the Axis
     */
    title?: string;
    /**
     * https://github.com/d3/d3-axis#axis_tickFormat
     */
    tickFormat?: Function;
    /**
     * https://github.com/d3/d3-axis#axis_ticks
     */
    ticks: any[];
}

/**
 * Represents an XAxis component
 * @return The X Axis component
 */
export function XAxis({
    position = "bottom",
    fields,
    scaleType,
    aggregate = false,
    showGridlines = true,
    title,
    tickSizeInner,
    tickSizeOuter,
    tickPadding,
    tickFormat,
    ticks,
    domain,
}: XAxisProps) {
    return (
        <React.Fragment>
            <Axis
                position={position}
                fields={fields}
                showGridlines={showGridlines}
                title={title}
                tickSizeInner={tickSizeInner}
                tickSizeOuter={tickSizeOuter}
                tickPadding={tickPadding}
                tickFormat={tickFormat}
                ticks={ticks}
            />
            <XScale fields={fields} scaleType={scaleType} aggregate={aggregate} domain={domain} fromAxis={true} />
        </React.Fragment>
    );
}
