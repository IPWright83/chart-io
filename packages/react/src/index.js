export * from "./components";
export * from "./hoc";
export * from "./hooks";
export * from "./store";
export * from "./types";
export * from "./utils";
export * from "./constants";
export * from "./themes";
