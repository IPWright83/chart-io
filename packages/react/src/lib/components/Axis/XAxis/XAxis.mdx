import { Meta, Story, Canvas } from "@storybook/blocks";
import LinkTo from "@storybook/addon-links/react";

<Meta title="Components/Axis/XAxis" />

# Axis Component

This `<XAxis>` component allows you to define a horizontal Axis on a chart. This is a utility component that combines both an <LinkTo kind="components-axis-types--string-axis#axis-component">`<Axis>`</LinkTo> component, along with an `<AutoScale>` component. It is recommended to use this component where possible to save duplicate configuration for an `<AutoScale>`.

Please see the <LinkTo kind="components-axis-types--string-axis#axis-component">Axis</LinkTo> component for more information on it's capabilities and configuration.

<Canvas>
    <Story id="components-axis-xaxis--x-axis-story" />
</Canvas>

### Props

| Prop            | Type                        | Default | Note                                                                                                                                                     |
| --------------- | --------------------------- | ------- | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `aggregate`     | `boolean`                   | false   | Whether the scale should be aggregated (used for stacked charts)                                                                                         |
| `domain`        | `Array<number/string/date>` | null    | An optional override of the domain for the scale. This will replace any auto calculated domain. See https://github.com/d3/d3-scale#continuous_domain     |
| **`fields`\***  | `Array<string>`             |         | The set of fields that this axis represents. These should correspond to the fields being displayed on this Axis within the data.                         |
| `position`      | `string`                    | bottom  | The position of the axis. One of `bottom` or `top`. This will use the appropriate [d3.axis](https://github.com/d3/d3-axis#axisTop) orientation function. |
| `scaleType`     | `string`                    | null    | An optional override of the type of scale that should be used. One of "linear", "time", "power", "log", "band", "point"                                  |
| `showGridlines` | `boolean`                   | true    | Determines whether gridlines should be shown against this axis.                                                                                          |
| `tickFormat`    | `function`                  | null    | See https://github.com/d3/d3-axis#axis_tickFormat                                                                                                        |
| `tickSizeInner` | `number`                    | 6       | See https://github.com/d3/d3-axis#axis_tickSizeInner.                                                                                                    |
| `tickSizeOuter` | `number`                    | 6       | See https://github.com/d3/d3-axis#axis_tickSizeOuter.                                                                                                    |
| `tickPadding`   | `number`                    | 3       | See https://github.com/d3/d3-axis#axis_tickPadding.                                                                                                      |
| `ticks`         | `any[]`                     | null    | See https://github.com/d3/d3-axis#axis_ticks                                                                                                             |
| `tickValues`    | `string[]`                  | null    | See https://github.com/d3/d3-axis#axis_tickValues                                                                                                        |
| `title`         | `string`                    | null    | The title of the axis. This will default to the value in field(s)                                                                                        |
