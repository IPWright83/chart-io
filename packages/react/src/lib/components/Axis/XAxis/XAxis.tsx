import type { IPosition, IScaleType, IValue } from "@chart-io/types";
import React from "react";

import { Axis, IAxisProps } from "../Axis";
import { XScale } from "../../Scale";

export interface IXAxisProps extends Omit<IAxisProps, "position"> {
    /**
     * The position of the axis [left, right]
     */
    position?: Omit<IPosition, "left" | "right">;
    /**
     * (Optional) An override of the domain to use with the d3 scale
     */
    domain?: IValue[];
    /**
     * (Optional) An override of the type of d3 scale touse
     */
    scaleType?: IScaleType;
    /**
     * Whether the underlying scale should aggregate the fields (e.g. for stacked charts)
     */
    aggregate?: boolean;
}

/**
 * Represents an XAxis component
 * @return The X Axis component
 */
export function XAxis({
    position = "bottom",
    fields,
    scaleType,
    aggregate = false,
    showGridlines = true,
    title,
    tickSizeInner,
    tickSizeOuter,
    tickPadding,
    tickFormat,
    ticks,
    tickValues,
    domain,
}: IXAxisProps) {
    return (
        <React.Fragment>
            <Axis
                position={position}
                fields={fields}
                showGridlines={showGridlines}
                title={title}
                tickSizeInner={tickSizeInner}
                tickSizeOuter={tickSizeOuter}
                tickPadding={tickPadding}
                tickFormat={tickFormat}
                ticks={ticks}
                tickValues={tickValues}
            />
            <XScale fields={fields} scaleType={scaleType} aggregate={aggregate} domain={domain} />
        </React.Fragment>
    );
}
