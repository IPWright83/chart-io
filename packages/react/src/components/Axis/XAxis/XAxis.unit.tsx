import * as d3 from "d3";
import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";

import { XAxis } from "./XAxis";

import { createStore, chartActions, chartSelectors } from "../../../store";

describe("XAxis", () => {
    const store = createStore();
    const scale = d3.scaleLinear().domain([0, 10]).range([0, 200]);

    store.dispatch(chartActions.setDimensions(200, 100, { left: 0, right: 0, top: 0, bottom: 0 }));
    store.dispatch(chartActions.setData([{ x: 0 }, { x: 5 }, { x: 10 }]));
    chartActions.setScales(["x"], scale, true)(store.dispatch);

    it("renders correctly", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <XAxis position="bottom" fields={["x"]} />
                </svg>
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("creates a scale in the store correctly", () => {
        const store = createStore();

        store.dispatch(chartActions.setDimensions(200, 100, { left: 0, right: 0, top: 0, bottom: 0 }));
        store.dispatch(chartActions.setData([{ x: 0 }, { x: 5 }, { x: 10 }]));

        render(
            <Provider store={store}>
                <svg>
                    <XAxis position="bottom" fields={["x"]} />
                </svg>
            </Provider>
        );

        const scale = chartSelectors.scales.getScale(store.getState(), "x");
        expect(scale).toBeDefined();
        expect(scale.domain()).toEqual([0, 10]);
        expect(scale.range()).toEqual([0, 200]);
    });
});
