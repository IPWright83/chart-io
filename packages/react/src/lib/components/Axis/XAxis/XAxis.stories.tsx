import React from "react";

import { Chart } from "../../Chart";
import { XAxis } from "./XAxis";

export default {
    title: "Components/Axis/XAxis",
    component: XAxis,
};

const XAxisTemplate = (args) => (
    <Chart {...args}>
        <XAxis fields={args.fields} />
    </Chart>
);

export const XAxisStory = XAxisTemplate.bind({});
XAxisStory.storyName = "X Axis";
XAxisStory.args = {
    data: [{ x: 0 }, { x: 10 }],
    fields: ["x"],
    height: 100,
    width: 800,
};
