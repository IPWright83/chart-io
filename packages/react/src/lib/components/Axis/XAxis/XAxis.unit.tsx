import { chartActions, chartSelectors, createStore } from "@chart-io/core";
import { d3 } from "@chart-io/core";

import { render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";

import { XAxis } from "./XAxis";

describe("XAxis", () => {
    const store = createStore();
    const scale = d3.scaleLinear().domain([0, 10]).range([0, 200]);

    store.dispatch(chartActions.setDimensions({
        width: 200,
        height: 100,
        margin: { left: 0, right: 0, top: 0, bottom: 0 }
    }));
    store.dispatch(chartActions.setChartData([{ x: 0 }, { x: 5 }, { x: 10 }]));
    store.dispatch(chartActions.setScales({ fields: ["x"], scale }));

    it("renders correctly", () => {
        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <XAxis position="bottom" fields={["x"]} />
                </svg>
            </Provider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("creates a scale in the store correctly", () => {
        const store = createStore();

        store.dispatch(chartActions.setDimensions({
            width: 200,
            height: 100,
            margin: { left: 0, right: 0, top: 0, bottom: 0 }
        }));
        store.dispatch(chartActions.setChartData([{ x: 0 }, { x: 5 }, { x: 10 }]));

        render(
            <Provider store={store}>
                <svg>
                    <XAxis position="bottom" fields={["x"]} />
                </svg>
            </Provider>,
        );

        const scale = chartSelectors.scales.getScale(store.getState(), "x", "plot");
        expect(scale).toBeDefined();
        expect(scale.domain()).toEqual([0, 10]);
        expect(scale.range()).toEqual([0, 200]);
    });
});
