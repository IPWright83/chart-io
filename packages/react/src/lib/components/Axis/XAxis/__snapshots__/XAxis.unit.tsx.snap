// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XAxis renders correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-axis"
    >
      <g
        transform="translate(0, 100)"
      >
        <g
          class="g-gridlines bottom"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          style="color: rgb(211, 211, 211); stroke-opacity: 0.8; stroke-width: 1; pointer-events: none;"
          text-anchor="middle"
        >
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(0.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(20.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(40.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(60.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(80.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(100.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(120.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(140.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(160.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(180.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(200.5,0)"
          >
            <line
              stroke="currentColor"
              y2="-100"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="3"
            />
          </g>
        </g>
        <g
          class="chart-it axis axis-bottom"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          style="user-select: none; color: rgb(51, 51, 51); stroke-opacity: 0.8; stroke-width: 1;"
          text-anchor="middle"
        >
          <path
            class="domain"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(0.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(20.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(40.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(60.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(80.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(100.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(120.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(140.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(160.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(180.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
          <g
            class="tick"
            opacity="0.000001"
            transform="translate(200.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            />
          </g>
        </g>
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
